html2/tokenizer_test.cpp

```diff
@@ -14,28 +14,128 @@
+                        StartTagToken{.tag_name = "html"s},
+                        CharacterToken{'\n'},
+                        EndTagToken{.tag_name = "html"s},
+                        CharacterToken{'\n'}});
     });
 
+    etest::test("comment, simple", [] {
+        auto tokens = run_tokenizer("<!-- Hello -->");
+
+        expect_eq(tokens, std::vector<Token>{CommentToken{.data = " Hello "}});
+    });
+
+    etest::test("comment, empty", [] {
+        auto tokens = run_tokenizer("<!---->");
+
+        expect_eq(tokens, std::vector<Token>{CommentToken{.data = ""}});
+    });
+
+    etest::test("comment, with dashes and bang", [] {
+        auto tokens = run_tokenizer("<!--!-->");
+
+        expect_eq(tokens, std::vector<Token>{CommentToken{.data = "!"}});
+    });
+
+    etest::test("comment, with new lines", [] {
+        auto tokens = run_tokenizer("<!--\nOne\nTwo\n-->");
+
+        expect_eq(tokens, std::vector<Token>{CommentToken{.data = "\nOne\nTwo\n"}});
+    });
+
+    etest::test("comment, multiple with new lines", [] {
+        auto tokens = run_tokenizer("<!--a-->\n<!--b-->\n<!--c-->");
+
+        expect_eq(tokens,
+                std::vector<Token>{CommentToken{.data = "a"},
+                        CharacterToken{'\n'},
+                        CommentToken{.data = "b"},
+                        CharacterToken{'\n'},
+                        CommentToken{.data = "c"}});
+    });
+
+    etest::test("comment, allowed to end with <!", [] {
+        auto tokens = run_tokenizer("<!--My favorite operators are > and <!-->");
+
+        expect_eq(tokens, std::vector<Token>{CommentToken{.data = "My favorite operators are > and <!"}});
+    });
+
+    etest::test("comment, nested comment", [] {
+        auto tokens = run_tokenizer("<!--<!---->");
+
+        expect_eq(tokens, std::vector<Token>{CommentToken{.data = "<!--"}});
+    });
+
+    etest::test("comment, nested comment closed", [] {
+        auto tokens = run_tokenizer("<!-- <!-- nested --> -->");
+
+        expect_eq(tokens,
+                std::vector<Token>{CommentToken{.data = " <!-- nested "},
+                        CharacterToken{' '},
+                        CharacterToken{'-'},
+                        CharacterToken{'-'},
+                        CharacterToken{'>'}});
+    });
+
+    etest::test("comment, abrupt closing in comment start", [] {
+        auto tokens = run_tokenizer("<!-->");
+
+        expect_eq(tokens, std::vector<Token>{CommentToken{.data = ""}});
+    });
+
+    etest::test("comment, abrupt closing in comment start dash", [] {
+        auto tokens = run_tokenizer("<!--->");
+
+        expect_eq(tokens, std::vector<Token>{CommentToken{.data = ""}});
+    });
+
+    etest::test("comment, incorrectly closed comment", [] {
+        auto tokens = run_tokenizer("<!--abc--!>");
+
+        expect_eq(tokens, std::vector<Token>{CommentToken{.data = "abc"}});
+    });
+
+    // TODO(mkiael): Enable when eof check is fixed
```

Comment:
What needs to be done with the eof check?

Reply:
I'm not sure, but it might be that the `!is_eof()` in the while condition makes the loop break to early, causing us to miss emitting eof tokens.

Reply:
Ah, yeah, you're probably right. Maybe we should be setting a flag if we encounter an eof-token in emit or something. :P